Kops Template YAML Formatting

gambol99 opened the pull request:
Adding an extra option to the toolbox templating to format the YAML before writing out; which is usefull to cleanup formating issues and as detecting errors in the template

- added a formating options --format-yaml to the toolbox template
- updated the cli documentation
- the option is off by default to retain behavior

gambol99 commented:
/assign @gambol99

gambol99 commented:
/assign @chrislovecnm

justinsb commented on cmd/kops/toolbox_template.go:
```diff
@@ -166,14 +163,39 @@ func runToolBoxTemplate(f *util.Factory, out io.Writer, options *toolboxTemplate
+		if len(rendered) <= 0 {
+			continue
+		}
 		io.WriteString(writer, rendered)
```
Probably better to call `writer.WriteString`, now that it is a bytes.Buffer (because we probably should have been checking for errors here previously, but we don't have to with bytes.Buffer, but we might as well call the method directly to trigger an error if we make this an io.Writer again!)  But not a big deal!

justinsb commented on cmd/kops/toolbox_template.go:
```diff
@@ -166,14 +163,39 @@ func runToolBoxTemplate(f *util.Factory, out io.Writer, options *toolboxTemplate
+
+	if options.formatYAML {
+		var data interface{}
+		err := yaml.Unmarshal(content, &data)
```
Is `---` actually part of yaml syntax?  i.e. does this work with more than one template?

If not, we might have to format each section independently..

justinsb commented on cmd/kops/toolbox_template.go:
```diff
@@ -166,14 +163,39 @@ func runToolBoxTemplate(f *util.Factory, out io.Writer, options *toolboxTemplate
+			return fmt.Errorf("encountered errors generating templates: %s", err)
+		}
+		if content, err = yaml.Marshal(data); err != nil {
+			return fmt.Errorf("encountered error formating the template: %s", err)
```
type: formatting

justinsb commented on cmd/kops/toolbox_template.go:
```diff
@@ -166,14 +163,39 @@ func runToolBoxTemplate(f *util.Factory, out io.Writer, options *toolboxTemplate
+			return fmt.Errorf("unable to open file: %s, error: %v", options.outputPath, err)
+		}
+		defer w.Close()
+		w.Write(content)
```
Should check for an error from w.Write here

justinsb commented on cmd/kops/toolbox_template.go:
```diff
@@ -166,14 +163,39 @@ func runToolBoxTemplate(f *util.Factory, out io.Writer, options *toolboxTemplate
+		defer w.Close()
+		w.Write(content)
+	} else {
+		out.Write(content)
```
Should check the error code here too

justinsb commented:
A few small things - want to see if you agree @gambol99 ?

gambol99 commented:
hi @justinsb

- fixed the spelling mistakes 
- amended the code to use the buffer writer than the io package and am checking the errors
- I've added a little --config-value which we use to grab merged configuration

justinsb commented:
/lgtm

Thanks @gambol99!

gambol99 commented:
damn! :'-( .. i just spotted a bug .. ok, raising a PR
